code/modules/mob/living/abilities/charge.dm

```diff
@@ -0,0 +1,388 @@
+
+	//Screenshake everyone near the impact site
+	for (var/mob/M in range(TP, obstacle))
+		shake_camera(M,10*TP,2)
+
+	obstacle.shake_animation(8*TP)
+	if (isliving(holder))
+		//Damage the charger and stun them
+		var/mob/living/L = holder
+		var/blocked = L.damaged_by_atom(obstacle, CHARGE_DAMAGE_BASE*TP + CHARGE_DAMAGE_DIST*tiles_moved)
+		L.apply_effect(3*TP, STUN, blocked)
+	stop()
+
+//Called when we reach max time or range
+//Drain the charger's stamina?
+/datum/extension/charge/proc/stop_peter_out()
+	stop()
+
+
+//We have ended the charge by successfully reaching our intended target. This is ideal
+/datum/extension/charge/proc/stop_success()
+	var/TP = get_total_power()
+	//Screenshake everyone near the impact site
+	for (var/mob/M in range(TP))
+		shake_camera(M,10*TP,2)
+	stop()
+
+//Extra procs
+//---------------------------
+
+///Called when this atom is hit by a charging mob
+//Return false if you want to stop the charge for some special reason.
+//Note that the reason of: "We are dense and not broken so you can't get past" is already checked for. It's gotta be something other than that
+/atom/proc/charge_act(var/atom/mover, var/power)
+	if (power > 0)
+		ex_act(max(4-power, 1)) //Ex act has tons of interactions already, we'll use it
+
+	return TRUE
+
+
+//Mobs take some damage and get stunned
+/mob/living/charge_act(var/atom/mover, var/power, var/dist)
+	shake_camera(src,10*power,1)
+	if (isliving(mover))
+		var/mob/living/L = mover
+		//We can't be hurt by things smaller than ourselves. they bounce off
+		if (L.mob_size < mob_size)
+			return FALSE
+
+	var/blocked = damaged_by_atom(mover, (CHARGE_DAMAGE_BASE*power) + (dist * CHARGE_DAMAGE_DIST))
+	apply_effect(3*power, STUN, blocked)
+	return TRUE
+
+
+//Called when this atom hits something while charging
+//target type will either be primary or secondary
+	//Primary = The thing our charge was aimed at
+	//Secondary = Something that got in the way while enroute to the primary target
+/atom/movable/proc/charge_impact(var/atom/obstacle, var/power, var/target_type, var/distance_travelled)
+	shake_camera(src,3,1)
+	return obstacle.charge_act(src, power, distance_travelled)
+
+
+//When a human does it, we call the same proc on their species. This allows various people to do stuff
+/mob/living/carbon/human/charge_impact(var/atom/obstacle, var/power, var/target_type, var/distance_travelled)
+	shake_camera(src,3,1)
+	if (species)
+		return species.charge_impact(src, obstacle, power, target_type, distance_travelled)
+	return ..()
+
+
+
+/datum/species/proc/charge_impact(var/mob/living/charger, var/atom/obstacle, var/power, var/target_type, var/distance_travelled)
+	return obstacle.charge_act(charger, power, distance_travelled)
+
+
+
+//	Triggering
+//------------------------
+/atom/movable/proc/charge_verb(var/atom/A)
+	set name = "Charge"
+	set category = "Abilities"
+
+
+	return charge_attack(A)
+
+
+/atom/movable/proc/can_charge(var/atom/target, var/error_messages = TRUE)
```

Comment:
Would you not want this on the extension itself? With a verb added by the extension when they're granted "charge"? We can probably assume that they're not going to be able to charge without said extension.

Reply:
No, its overrideable so that various mob types can do their own checks. There's a human version just a few lines down from that